public/app/core/components/RolePicker/RolePickerMenu.tsx

```diff
@@ -0,0 +1,585 @@
+import React, { FormEvent, useCallback, useEffect, useState } from 'react';
+import { css, cx } from '@emotion/css';
+import {
+  Button,
+  Checkbox,
+  CustomScrollbar,
+  HorizontalGroup,
+  RadioButtonGroup,
+  Tooltip,
+  useStyles2,
+  useTheme2,
+} from '@grafana/ui';
+import { GrafanaTheme2, SelectableValue } from '@grafana/data';
+import { getSelectStyles } from '@grafana/ui/src/components/Select/getSelectStyles';
+import { OrgRole, Role } from 'app/types';
+
+type BuiltInRoles = { [key: string]: Role[] };
+
+const BuiltinRoles = ['Viewer', 'Editor', 'Admin'];
+const BuiltinRoleOption: Array<SelectableValue<OrgRole>> = BuiltinRoles.map((r: string) => ({
+  label: r,
+  value: r as OrgRole,
+}));
+
+const fixedRoleGroupNames: { [key: string]: string } = {
+  ldap: 'LDAP',
+  current: 'Current org',
+};
+
+interface RolePickerMenuProps {
+  builtInRole: OrgRole;
+  builtInRoles: BuiltInRoles;
+  options: Role[];
+  appliedRoles: Role[];
+  showGroups?: boolean;
+  onSelect: (roles: Role[]) => void;
+  onBuiltInRoleSelect?: (role: OrgRole) => void;
+  onUpdate: (newBuiltInRole: OrgRole, newRoles: string[]) => void;
+  onClear?: () => void;
+}
+
+export const RolePickerMenu = ({
+  builtInRole,
+  builtInRoles,
+  options,
+  appliedRoles,
+  showGroups,
+  onSelect,
+  onBuiltInRoleSelect,
+  onUpdate,
+  onClear,
+}: RolePickerMenuProps): JSX.Element => {
+  const [selectedOptions, setSelectedOptions] = useState<Role[]>(appliedRoles);
+  const [selectedBuiltInRole, setSelectedBuiltInRole] = useState<OrgRole>(builtInRole);
+  const [selectedBuiltInRoles, setSelectedBuiltInRoles] = useState<Role[]>(builtInRoles[builtInRole]);
+  const [showSubMenu, setShowSubMenu] = useState(false);
+  const [openedMenuGroup, setOpenedMenuGroup] = useState('');
+  const [subMenuOptions, setSubMenuOptions] = useState<Role[]>([]);
+
+  const theme = useTheme2();
+  const styles = getSelectStyles(theme);
+  const customStyles = useStyles2(getStyles);
+
+  const getOptionGroups = useCallback(() => {
+    const groupsMap: { [key: string]: Role[] } = {};
+    options.forEach((role) => {
+      if (role.name.startsWith('fixed:')) {
+        const groupName = getRoleGroup(role);
+        if (groupsMap[groupName]) {
+          groupsMap[groupName].push(role);
+        } else {
+          groupsMap[groupName] = [role];
+        }
+      }
+    });
+
+    const groups = [];
+    for (const groupName in groupsMap) {
+      if (Object.prototype.hasOwnProperty.call(groupsMap, groupName)) {
```

Comment:
```suggestion
    for (const groupName of Object.keys(groupsMap)) {
```
Avoid using `for .. in`, `for .. of` is a better alternative as it iterates only over own properties. 

However, i'd recommend using `map` here:
```typescript
return Object.entries(groupsMap)
      .map(([groupName, roles]) => {
        return {
          name: fixedRoleGroupNames[groupName] || capitalize(groupName),
          value: groupName,
          options: roles.sort(sortRolesByName),
        };
      })
      .sort((a, b) => (a.name < b.name ? -1 : 1));
```
In general I'd prefer using arrays over objects as their manipulation is easier in JS.